### PR TITLE
New CANDataStructureGenerator powered by CANMessage class

### PR DESCRIPTION
**Pull request by fgiaimo**
To be merged together with newly formatted .can files in higher software layers to avoid parsing errors

### CODE REVIEW

File: AutomotiveData.odvd
Comment (chrberger) [inline AutomotiveData.odvd:167]: okay to remove as discussed

File: AutomotiveData.odvd
Comment (chrberger) [inline AutomotiveData.odvd:175]: okay to remove as discussed

File: CMakeLists.txt
Comment (chrberger) [inline CMakeLists.txt:445]: okay to remove as discussed

File: automotive/CANMessageMapping.can
Comment (chrberger) [inline automotive/CANMessageMapping.can:1]: okay to remove as discussed

File: automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/CANDataModel.xtext
Comment (chrberger) [inline automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/CANDataModel.xtext]: @fgiaimo Could you please improve the language design slightly (cf. OpenDLV.core):

"CAN Message myMessage : 8 {"
--> 
"CANMessage myMessage is 8 bytes {

Then, the content is more intuitive at the first glance.

File: automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend
Comment (chrberger) [inline automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend:42]: Identation?

File: automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend
Comment (chrberger) [inline automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend]: Please remove unneeded lines.

File: automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend
Comment (chrberger) [inline automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend]: @fgiaimo could you please check with @HangYin2015 if the UPPAAL part from the generator could be removed entirely to improve readability?

File: automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend
Comment (chrberger) [inline automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend:421]: identation

File: automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend
Comment (chrberger) [inline automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend:514]: indentation

File: automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend
Comment (chrberger) [inline automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend:549]: indentation

File: automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend
Comment (chrberger) [inline automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend:710]: @fgiaimo What does this comment mean?

File: automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend
Comment (chrberger) [inline automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend:834]: indentation

File: automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend
Comment (chrberger) [inline automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend]: @fgiaimo please check with @HangYin2015 if we could remove UPPAAL entirely

File: automotive/odcantools/libodcantools/include/CANMessage.h
Comment (chrberger) [inline automotive/odcantools/libodcantools/include/CANMessage.h]: `use namespace std;`

File: automotive/odcantools/libodcantools/include/CANMessage.h
Comment (chrberger) [inline automotive/odcantools/libodcantools/include/CANMessage.h:5]: `#include <string>`

File: automotive/odcantools/libodcantools/include/CANMessage.h
Comment (chrberger) [inline automotive/odcantools/libodcantools/include/CANMessage.h]: @fgiaimo Can we have `const Type&` here to avoid unnecessary value copying (CAN message handling is typically executed with higher frequencies and saving costs by avoiding call by value is preferred)

File: automotive/odcantools/libodcantools/include/CANMessage.h
Comment (chrberger) [inline automotive/odcantools/libodcantools/include/CANMessage.h]: same here

File: automotive/odcantools/libodcantools/include/CANMessage.h
Comment (chrberger) [inline automotive/odcantools/libodcantools/include/CANMessage.h]: and here

File: automotive/odcantools/libodcantools/include/CANMessage.h
Comment (chrberger) [inline automotive/odcantools/libodcantools/include/CANMessage.h]: and here

File: automotive/odcantools/libodcantools/include/CANMessage.h
Comment (chrberger) [inline automotive/odcantools/libodcantools/include/CANMessage.h]: and here

File: automotive/odcantools/libodcantools/src/CANMessage.cpp
Comment (chrberger) [inline automotive/odcantools/libodcantools/src/CANMessage.cpp:15]: We should preallocate 8 bytes in our vectors to avoid vector copying back and forth at runtime.

Please check if a preallocated vector does not collide with your logic (i.e., when you check for payload.empty() for instance)

File: automotive/odcantools/libodcantools/src/CANMessage.cpp
Comment (chrberger) [inline automotive/odcantools/libodcantools/src/CANMessage.cpp]: The `string::compare`might be expensive --> would there be a better implementation?

File: automotive/odcantools/libodcantools/src/CANMessage.cpp
Comment (chrberger) [inline automotive/odcantools/libodcantools/src/CANMessage.cpp]: Perhaps, an `enum` is better to avoid expensive string comparisons

File: automotive/odcantools/libodcantools/src/CANMessage.cpp
Comment (chrberger) [inline automotive/odcantools/libodcantools/src/CANMessage.cpp]: Perhaps, an `enum` is better to avoid expensive string comparisons

**Review by chrberger — CHANGES_REQUESTED** [review]
@fgiaimo Thanks for the new implementation; just a few minor changes are required from my end.

File: automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend
Comment (fgiaimo) [reply automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend:42]: The indentation is imperfect to minimize the number of warnings (ironically, the warnings say "Inconsistent indentation") returned by xtend compilation. 
It is unclear to me why is it so, but if I follow the "cleaner" indentation that we like, each of these weird-indented lines will result in a new warning when the generator is compiled.
If this is an acceptable outcome, I will fix them.

File: automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend
Comment (fgiaimo) [reply automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend]: Done and removed

File: automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend
Comment (fgiaimo) [reply automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend:710]: It is related to https://github.com/se-research/OpenDaVINCI/blob/2017Q1.refactoring.cangenerator/automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend#L763 , so if a mapping is defined over a set of several physical CAN IDs, the encode will return only the first CAN message of the set (the decode still works as expected, though).
The multiple-CAN encode has never been used/needed so far.

File: automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend
Comment (chrberger) [reply automotive/odCANDataStructureGenerator/org.opendavinci.candatamodel/src/org/opendavinci/generator/CANDataModelGenerator.xtend:42]: If this indentation fixes the warnings, let's stick to it.

File: automotive/odcantools/libodcantools/include/CANMessage.h
Comment (fgiaimo) [reply automotive/odcantools/libodcantools/include/CANMessage.h]: Why do we need `std`?

**Comment by chrberger** [comment]
This header file does not know this namespace for the fields that you use below. The using namespace line will prevent some Jenkins instances fail.

File: automotive/odcantools/libodcantools/src/CANMessage.cpp
Comment (fgiaimo) [reply automotive/odcantools/libodcantools/src/CANMessage.cpp:15]: Reminder to myself: use `void vector::reserve (size_type n);`

**Comment by fgiaimo** [comment]
Hi @chrberger, can you confirm that the PR addresses now all your feedback? 
I am building the docker images of ODV, later I'll build the ones for ODLV.core and COPPLAR.core. 
If in the end everything's working we could merge into pre-release.

**Review by chrberger — APPROVED** [review]
@fgiaimo Please either re-base or merge pre-release into your branch. Then, I suggest that we test your branch first on the vehicle before we go full-way and integrate it with master. What do you think?

**Comment by fgiaimo** [comment]
Yes, I merged pre-release already yesterday. When my compilations are green I will let you know and after the vehicle tests we can merge the PR in pre-release and then in master

**Comment by chrberger** [comment]
Yes, great. Thanks.

What test(s) shall we conduct tomorrow at AstaZero?

**Comment by fgiaimo** [comment]
The ps3 controller on both vehicles should be enough

**Comment by chrberger** [comment]
Yes, I was assuming. So, there is no further manual step in-between and we can just use the temporary Docker image from your branch to build for Rhino and Snowfox and run the PS3Controller?

**Comment by fgiaimo** [comment]
I assume yes. However the opendlv.core & copplar.core images have to contain the correct .can file. I can send the details of which branches to build when I have green light. Besides I still need to adjust to the new software layer structure that we have in copplar

**Pull request merged**
